prql-compiler/src/prql.pest

```diff
@@ -11,30 +11,21 @@ WHITESPACE = _{ " " | "\t" }
 // Need to exclude # in strings (and maybe confirm whether this the syntax we want)
 COMMENT = _{ "#" ~ (!NEWLINE ~ ANY) * }
 
-query = { SOI ~ NEWLINE* ~ (query_def ~ NEWLINE+)? ~ (( func_def | table | pipeline ) ~ ( NEWLINE+ | &EOI ))* ~ EOI }
+query = { SOI ~ NEWLINE* ~ query_def ~ NEWLINE* ~ (( func_def | table | pipeline ) ~ ( NEWLINE+ | &EOI ))* ~ EOI }
 
-query_def = { WHITESPACE* ~ "prql" ~ named_expr_simple*  }
+query_def = { WHITESPACE* ~ ("prql" ~ named_expr* ~ NEWLINE)? }
 
-func_def = { "func" ~ ident ~ func_def_params ~ "=" ~ expr }
-// TODO: we could force the named parameters to follow the positional ones here.
-// But pest requires some special handling because if `ident` is before
-// `named_arg`, then `x:y` will be parsed as `x` and then error on `:y`.
-// Possibly we can mandate whitespace to follow, but pest has annoying magic
-// which removes whitespace in non-atomic rules, so we'd need to explore that.
-func_def_params = { ( named_term_simple | ident )* }
-table = { "table" ~ ident ~ "=" ~ "(" ~ NEWLINE* ~ pipeline ~ NEWLINE* ~ ")" }
+func_def = { "func" ~ func_def_flags ~ ident ~ func_def_params ~ "=" ~ expr }
 
-pipe = _{ NEWLINE+ | "|" }
-pipeline = { transformation ~ (pipe ~ transformation)* }
+func_def_flags = @{ ("transform" | "aggregation" | "window" | "")? }
```

Comment:
Should we have some sort of syntax for these "type" arguments? For the moment I'll suggest a TODO

```suggestion
// TODO: consider some syntax for these func "type" args; currently they're only 
// differentiated from params by their position
func_def_flags = @{ ("transform" | "aggregation" | "window" | "")? }
```

Reply:
Are there any examples of how these work? I see these as `FuncKind`s, but I couldn't see any practical examples. I'm presuming they're for giving the compile information about what the function is doing?